File: educational-core/src/com/jetbrains/edu/coursecreator/ui/CCCreateStudyItemDialog.java
package com.jetbrains.edu.coursecreator.ui;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.openapi.ui.InputValidatorEx;
import com.intellij.openapi.ui.ValidationInfo;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.ui.DocumentAdapter;
import com.jetbrains.edu.coursecreator.CCUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import javax.swing.event.DocumentEvent;

public class CCCreateStudyItemDialog extends DialogWrapper {
  private final CCCreateStudyItemPanel myPanel;
  private final InputValidatorEx myValidator;

  public CCCreateStudyItemDialog(@Nullable Project project,
                                 String itemName,
                                 String thresholdName,
                                 int thresholdIndex,
                                 @NotNull VirtualFile parent) {
    super(project);
    myPanel = new CCCreateStudyItemPanel(itemName, thresholdName, thresholdIndex);
    setTitle("Create New " + StringUtil.toTitleCase(itemName));
    init();
    myValidator = new CCUtils.PathInputValidator(parent);
    myPanel.getNameField().getDocument().addDocumentListener(new DocumentAdapter() {
      @Override
      protected void textChanged(DocumentEvent e) {
        initValidation();
      }
    });
  }

  @Nullable
  @Override
  protected JComponent createCenterPanel() {
   return myPanel;
  }

  public String getName() {
    return myPanel.getItemName();
  }

  public int getIndexDelta() {
    return myPanel.getIndexDelta();
  }

  @Nullable
  @Override
  protected ValidationInfo doValidate() {
    String name = myPanel.getItemName();
    if (name == null) {
      return new ValidationInfo("empty name");
    }
    if (myValidator.checkInput(name)) {
      return null;
    }
    String errorText = myValidator.getErrorText(name);
    return errorText == null ? null : new ValidationInfo(errorText);
  }
}

File: educational-core/src/com/jetbrains/edu/coursecreator/ui/CCCreateStudyItemDialog.kt
package com.jetbrains.edu.coursecreator.ui

import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.InputValidatorEx
import com.intellij.ui.components.JBTextField
import com.intellij.ui.layout.*
import com.jetbrains.edu.coursecreator.CCStudyItemPathInputValidator
import com.jetbrains.edu.coursecreator.actions.studyItem.NewStudyItemInfo
import com.jetbrains.edu.coursecreator.actions.studyItem.NewStudyItemUiModel
import com.jetbrains.edu.coursecreator.createItemTitleMessage
import com.jetbrains.edu.coursecreator.ui.CCItemPositionPanel.Companion.AFTER_DELTA
import com.jetbrains.edu.learning.courseFormat.Course
import com.jetbrains.edu.learning.messages.EduCoreBundle
import org.jetbrains.annotations.Nls
import javax.swing.JComponent

abstract class CCCreateStudyItemDialogBase(
  project: Project,
  course: Course,
  protected val model: NewStudyItemUiModel
) : CCDialogWrapperBase(project) {

  private val nameField: JBTextField = JBTextField(model.suggestedName, TEXT_FIELD_COLUMNS)
  private val validator: InputValidatorEx = CCStudyItemPathInputValidator(project, course, model.itemType, model.parentDir)


  init {
    title = model.itemType.createItemTitleMessage
  }

  override fun postponeValidation(): Boolean = false

  override fun createCenterPanel(): JComponent {
    addTextValidator(nameField) { it.getValidatorText() }
    return panel {
      if (showNameField()) {
        row("${EduCoreBundle.message("course.creator.new.study.item.label.name")}:") { nameField() }
      }
      createAdditionalFields(this)
    }
  }

  @Nls
  private fun String?.getValidatorText() =
    when {
      isNullOrEmpty() -> EduCoreBundle.message("course.creator.new.study.item.empty.name")
      !validator.checkInput(this) -> validator.getErrorText(this)
      else -> null
    }

  override fun getPreferredFocusedComponent(): JComponent? = nameField

  open fun showAndGetResult(): NewStudyItemInfo? =
    if (showAndGet()) createNewStudyItemInfo() else null

  protected open fun createNewStudyItemInfo(): NewStudyItemInfo {
    return NewStudyItemInfo(
      nameField.text,
      model.baseIndex + AFTER_DELTA,
      model.studyItemVariants.first().producer
    )
  }

  protected open fun createAdditionalFields(builder: LayoutBuilder) {}
  protected open fun showNameField(): Boolean = true

  companion object {
    const val TEXT_FIELD_COLUMNS: Int = 30
  }
}
